### ISSUE TITLE
Why is the mask deduced by sam3 in such a chaotic state?

### ISSUE THREAD

**Issue by Lxycherryup** [facebookresearch/sam3#512]
<img width="474" height="608" alt="Image" src="https://github.com/user-attachments/assets/8261318c-a416-4a03-926e-59718da2c344" />

# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# SAM 3 Image Segmentation Demo
# 支持文本提示和视觉提示（边界框）的图像分割

import os

import matplotlib
matplotlib.use("Agg")  # headless server — no display required
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np
import torch
from PIL import Image
from matplotlib.colors import to_rgb

import sam3
from sam3 import build_sam3_image_model
from sam3.model.box_ops import box_xywh_to_cxcywh
from sam3.model.sam3_image_processor import Sam3Processor
from sam3.visualization_utils import draw_box_on_image, normalize_bbox

# ──────────────────────────────────────────────
# 全局设置
# ──────────────────────────────────────────────

torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.allow_tf32 = True
torch.autocast("cuda", dtype=torch.bfloat16).__enter__()

sam3_root = os.path.join(os.path.dirname(sam3.__file__), "..")

# ──────────────────────────────────────────────
# 可视化辅助
# ──────────────────────────────────────────────

_VIS_COLORS = [
    "#e63232", "#28d928", "#3366ff",
    "#ffcc00", "#cc33ff", "#00e5e5",
    "#ff8000", "#80ff00", "#0080ff",
]

# 填充透明度（降低以避免整图染色）
_FILL_ALPHA = 0.30
# 轮廓线宽度（像素）
_CONTOUR_WIDTH = 3
_CONTOUR_ALPHA = 0.85

def _mask_contour(mask_bool, width=3):
    """提取 mask 的边界轮廓（宽度为 width 像素）。
    用简单的腐蚀实现，无需 scipy/cv2。"""
    h, w = mask_bool.shape
    pad = width
    padded = np.pad(mask_bool, pad, mode="constant", constant_values=False)

    # kernel 范围内全 True → 腐蚀后保留
    eroded = np.ones_like(mask_bool)
    for dy in range(-pad, pad + 1):
        for dx in range(-pad, pad + 1):
            eroded &= padded[pad + dy : pad + dy + h, pad + dx : pad + dx + w]

    # 轮廓 = 原 mask − 腐蚀
    return mask_bool & (~eroded)

def _plot_results_composite(img, results, prompt=""):
    """修复版可视化：加权混合 + 轮廓线。

    原始 bug 分析
    ─────────────
    旧代码对 composite 做 **赋值**（=），后绘制的 mask 直接覆盖前面的颜色；
    但更致命的是，SAM3 输出的多个候选 mask 往往 **高度重叠**，几乎每个像素
    都属于至少一个 mask → 以 alpha=0.45 叠加到底图上，整张图被均匀染色。

    修复策略
    ────────
    1. 累加混合：用 color_acc / count 取均值色，重叠区不会越叠越亮
    2. 降低填充 alpha（0.30），底图始终清晰可见
    3. 增加每个 mask 的 **轮廓线**（高 alpha），让目标边界一目了然
    """
    masks  = results["masks"]   # (N, 1, H, W) bool
    boxes  = results["boxes"]   # (N, 4) XYXY abs
    scores = results["scores"]  # (N,)
    n = len(scores)

    img_arr = np.array(img)
    h_img, w_img = img_arr.shape[:2]

    fig, ax = plt.subplots(figsize=(12, 8))
    ax.imshow(img_arr)

    if n > 0:
        # ── 1. 填充层：累加取均值，避免覆盖 ──
        color_acc = np.zeros((h_img, w_img, 3), dtype=np.float64)
        count     = np.zeros((h_img, w_img),    dtype=np.float64)

        for i in range(n):
            m = masks[i].squeeze(0).cpu().numpy().astype(bool)
            c = np.array(to_rgb(_VIS_COLORS[i % len(_VIS_COLORS)]))
            color_acc[m] += c
            count[m]     += 1.0

        has_mask = count > 0
        fill_layer = np.zeros((h_img, w_img, 4), dtype=np.float32)
        fill_layer[has_mask, :3] = (color_acc[has_mask] / count[has_mask, None]).astype(np.float32)
        fill_layer[has_mask, 3]  = _FILL_ALPHA

        ax.imshow(fill_layer, interpolation="nearest")

        # ── 2. 轮廓层：每个 mask 独立画边界 ──
        contour_layer = np.zeros((h_img, w_img, 4), dtype=np.float32)
        for i in range(n):
            m = masks[i].squeeze(0).cpu().numpy().astype(bool)
            edge = _mask_contour(m, width=_CONTOUR_WIDTH)
            c = np.array(to_rgb(_VIS_COLORS[i % len(_VIS_COLORS)]), dtype=np.float32)
            contour_layer[edge, :3] = c
            contour_layer[edge, 3]  = _CONTOUR_ALPHA

        ax.imshow(contour_layer, interpolation="nearest")

        # ── 3. 检测框 + 标签 ──
        for i in range(n):
            c = to_rgb(_VIS_COLORS[i % len(_VIS_COLORS)])
            b = boxes[i].cpu().numpy()
            x1, y1, x2, y2 = b
            rect = mpatches.Rectangle(
                (x1, y1), x2 - x1, y2 - y1,
                linewidth=2, edgecolor=c, facecolor="none",
                linestyle="--", zorder=5,
            )
            ax.add_patch(rect)
            prob = scores[i].item()
            label = f"{prompt}  {prob:.2f}" if prompt else f"id={i}  {prob:.2f}"
            ax.text(
                x1 + 3, max(y1 - 6, 2), label,
                color="white", fontsize=8, fontweight="bold", va="bottom", zorder=6,
                bbox=dict(boxstyle="round,pad=0.25", facecolor=c,
                          alpha=0.85, edgecolor="none"),
            )

    print(f"found {n} object(s)")
    ax.axis("off")
    return fig, ax

# ── 用户配置 ──────────────────────────────────
CHECKPOINT_PATH = "/root/autodl-tmp/sam3-main/sam3_logs/vine/checkpoints/checkpoint.pt"
IMAGE_PATH      = "/root/autodl-tmp/data1/val/IMG20250103120658.jpg"
PROMPT          = "vine"
OUTPUT_DIR      = "/root/autodl-tmp/sam3-main/demo_results"
# ──────────────────────────────────────────────

# ──────────────────────────────────────────────
# 构建模型
# ──────────────────────────────────────────────

def build_model(checkpoint_path=CHECKPOINT_PATH):
    bpe_path = os.path.join(sam3_root, "assets", "bpe_simple_vocab_16e6.txt.gz")
    model = build_sam3_image_model(
        bpe_path=bpe_path,
        checkpoint_path=None,
        load_from_HF=False,
        eval_mode=True,
    )
    ckpt = torch.load(checkpoint_path, map_location="cpu")
    if isinstance(ckpt, dict) and "model" in ckpt:
        print(f"[INFO] Trainer checkpoint (epoch={ckpt.get('epoch', '?')})")
        missing, _ = model.load_state_dict(ckpt["model"], strict=False)
        if missing:
            print(f"[WARN] {len(missing)} missing keys")
    else:
        model.load_state_dict(ckpt, strict=False)
    return model

# ──────────────────────────────────────────────
# 文本提示分割
# ──────────────────────────────────────────────

def demo_text_prompt(processor, inference_state, image_path, prompt=PROMPT, save_path=None):
    """使用自然语言描述来分割目标物体"""
    processor.reset_all_prompts(inference_state)
    inference_state = processor.set_text_prompt(state=inference_state, prompt=prompt)

    img = Image.open(image_path)
    fig, ax = _plot_results_composite(img, inference_state, prompt=prompt)
    ax.set_title(f'prompt: "{prompt}"', fontsize=12)
    if save_path:
        os.makedirs(os.path.dirname(save_path), exist_ok=True)
        fig.savefig(save_path, dpi=150, bbox_inches="tight")
        print(f"  Saved → {save_path}")
    plt.close(fig)

# ──────────────────────────────────────────────
# 单框视觉提示分割
# ──────────────────────────────────────────────

def demo_single_box_prompt(processor, inference_state, image_path, width, height,
                           save_dir=None):
    """使用单个边界框作为视觉提示"""
    box_input_xywh = torch.tensor([480.0, 290.0, 110.0, 360.0]).view(-1, 4)
    box_input_cxcywh = box_xywh_to_cxcywh(box_input_xywh)

    norm_box_cxcywh = normalize_bbox(box_input_cxcywh, width, height).flatten().tolist()
    print("Normalized box input:", norm_box_cxcywh)

    processor.reset_all_prompts(inference_state)
    inference_state = processor.add_geometric_prompt(
        state=inference_state, box=norm_box_cxcywh, label=True
    )

    img = Image.open(image_path)
    image_with_box = draw_box_on_image(img, box_input_xywh.flatten().tolist())
    plt.imshow(image_with_box)
    plt.axis("off")
    if save_dir:
        plt.savefig(os.path.join(save_dir, "single_box_input.png"), dpi=150, bbox_inches="tight")
        print(f"  Saved → {os.path.join(save_dir, 'single_box_input.png')}")
    plt.close()

    img = Image.open(image_path)
    fig, _ = _plot_results_composite(img, inference_state)
    if save_dir:
        fig.savefig(os.path.join(save_dir, "single_box_result.png"), dpi=150, bbox_inches="tight")
        print(f"  Saved → {os.path.join(save_dir, 'single_box_result.png')}")
    plt.close(fig)

# ──────────────────────────────────────────────
# 多框视觉提示分割（正负样本）
# ──────────────────────────────────────────────

def demo_multi_box_prompt(processor, inference_state, image_path, width, height,
                          save_dir=None):
    """使用多个边界框（正/负样本）作为视觉提示"""
    box_input_xywh = [[480.0, 290.0, 110.0, 360.0], [370.0, 280.0, 115.0, 375.0]]
    box_input_cxcywh = box_xywh_to_cxcywh(torch.tensor(box_input_xywh).view(-1, 4))
    norm_boxes_cxcywh = normalize_bbox(box_input_cxcywh, width, height).tolist()

    box_labels = [True, False]

    processor.reset_all_prompts(inference_state)
    for box, label in zip(norm_boxes_cxcywh, box_labels):
        inference_state = processor.add_geometric_prompt(
            state=inference_state, box=box, label=label
        )

    img = Image.open(image_path)
    image_with_box = img
    for i in range(len(box_input_xywh)):
        color = (0, 255, 0) if box_labels[i] else (255, 0, 0)
        image_with_box = draw_box_on_image(image_with_box, box_input_xywh[i], color)
    plt.imshow(image_with_box)
    plt.axis("off")
    if save_dir:
        plt.savefig(os.path.join(save_dir, "multi_box_input.png"), dpi=150, bbox_inches="tight")
        print(f"  Saved → {os.path.join(save_dir, 'multi_box_input.png')}")
    plt.close()

    img = Image.open(image_path)
    fig, _ = _plot_results_composite(img, inference_state)
    if save_dir:
        fig.savefig(os.path.join(save_dir, "multi_box_result.png"), dpi=150, bbox_inches="tight")
        print(f"  Saved → {os.path.join(save_dir, 'multi_box_result.png')}")
    plt.close(fig)

# ──────────────────────────────────────────────
# 主函数
# ──────────────────────────────────────────────

def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    print(f"[INFO] Loading checkpoint: {CHECKPOINT_PATH}")
    model = build_model(CHECKPOINT_PATH)
    print("[INFO] Model ready.\n")

    image = Image.open(IMAGE_PATH)
    width, height = image.size
    print(f"[INFO] Image: {IMAGE_PATH}  ({width}x{height})\n")

    processor = Sam3Processor(model, confidence_threshold=0.5)
    with torch.autocast("cuda", dtype=torch.bfloat16):
        inference_state = processor.set_image(image)

    print("=" * 50)
    print(f"Demo 1: 文本提示分割 (prompt='{PROMPT}')")
    print("=" * 50)
    with torch.autocast("cuda", dtype=torch.bfloat16):
        demo_text_prompt(
            processor, inference_state, IMAGE_PATH, prompt=PROMPT,
            save_path=os.path.join(OUTPUT_DIR, "demo1_text_prompt.png"),
        )

    print("=" * 50)
    print("Demo 2: 单框视觉提示分割")
    print("=" * 50)
    with torch.autocast("cuda", dtype=torch.bfloat16):
        demo_single_box_prompt(
            processor, inference_state, IMAGE_PATH, width, height,
            save_dir=OUTPUT_DIR,
        )

    print("=" * 50)
    print("Demo 3: 多框视觉提示分割（正/负样本）")
    print("=" * 50)
    with torch.autocast("cuda", dtype=torch.bfloat16):
        demo_multi_box_prompt(
            processor, inference_state, IMAGE_PATH, width, height,
            save_dir=OUTPUT_DIR,
        )

    print(f"\n[Done] 结果保存至: {OUTPUT_DIR}")

if __name__ == "__main__":
    main()
